### ISSUE TITLE
[Suggestion] Use platform independent library to handle crypto ?

### ISSUE THREAD

**Issue by tuxuser** [emoose/xvdtool#6]
**Current situation**
- AES / RC4 is handled by custom implementation to LibXboxOne.
- SHA256 is handled by .NET's System.Security.Cryptography
- RSA(3) is handled by Windows-exclusive NCrypt P/Invoke (NCrypt)

**Proposal**
Implement cryptography in a unified fashion, either **System.Security.Cryptography** or **BouncyCastle** ?!

**Result**
- Multi platform compatibility
- More failsafe by using widely accepted / distributed library
- Better maintainability

**Comment by emoose** [comment]
Well I'd be happy to replace them with better supported code, as long as they don't interfere with how xvdtool currently works anyway. There were some reasons for the way it's setup now though, but hopefully time has solved them by now:

#### Compatibility

The AES stuff was because they use some weird variant, I think it's called AES-CTR but not 100% sure. I spent ages trying to find something to use for it, but just ended up finding some code that was easy to customize.

You can see the main change here with CryptData calling MorphIv between blocks: https://github.com/emoose/xvdtool/blob/249b563961d4a3d5157055986253e64b54dea1fc/LibXboxOne/Shared.cs#L391
(looking at it now though, I think normal AES-ECB with that MorphIv and the IV XOR on L424 should work fine, might be worth experimenting with... guess I might have thought more AES changes would be needed, but IIRC it should just be that IV stuff)

RC4 was down to this too, had problems with incompatible RC4 code in the past so I just used something I knew worked fine, if there's something we can slot-in that still works with everything I'd be happy to use it though.

#### Self-containment

Back then I really wanted xvdtool to be mostly self-contained, IIRC NuGet wasn't as popular then and I don't think there was even support for it outside of Windows, so I figured keeping dependencies low was the best action.

Hopefully that's changed by now though, I know MS have been making .NET stuff more open lately, if that means we can use outside libs without any hassle for non-Windows users/devs then I guess we can rule this out.

#### Laziness

The RSA P/Invoke mess was really just me being lazy, that was probably the last thing I'd worked on, and at that point really I just wanted to get it all working without bothering with different key formats and all that stuff. :P

Also found MS docs are really vague when it comes to B/NCrypt things, for all I know .NET probably could have supported it all natively, I just found it easier to P/Invoke than spend days combing through MS knowledge base tomes...

Of course it'd be great if we can replace that with native .NET instead, though like I said above it shouldn't interfere with the current way xvdtool works (meaning it should still work with raw RSA3 keys, so that testing extracted keys is as simple as possible, etc)

(also btw thanks for the p/rs you've sent, have been meaning to get back to you on them, will have to give them a look over properly this weekend)

**Comment by tuxuser** [comment]
Thanks for your detailed response.

1) I figured that out-of-spec usage of AES was the main reason to switch to an independent Aes implementation.

2) I am already working on handling it all by BouncyCastle (incl. raw BCryptBlob import of RSA keys). It will be all unit-test covered, so you know when something breaks.

And indeed, .NET Core incl. the option to use NuGet packages without a hassle is pretty sweet, been using it for quite a while now ;)

**Comment by emoose** [comment]
Sounds good, looking forward to it!

I did some digging and it seems it's actually AES-XTS being used, at least the MultiplyTweak function here https://github.com/sgan81/apfs-fuse/blob/8e7fbb4451f599eb9be77a5835ea55e95b738afd/ApfsLib/AesXts.cpp#L85 looks the same as the MorphIv function I'd reversed, would make sense too since XTS is apparently tailored for disk encryption.

Not sure if there's any C# impl. that works exactly like MS's does but it's probably worth looking out for, if not we should probably at least get rid of the custom AES stuff and use .NETs AES instead. (I did see https://gist.github.com/hanswolff/8809275 which seemed like a nice "proper" way of implementing AES-CTR on top of .NETs AES, probably wouldn't be too hard to adapt to XTS)

**Comment by tuxuser** [comment]
Funny, I was looking at it for at least 3 hours now, trying various things - AES CCM, AES SIV etc. - shame on me.
Can you contact me on discord or somewhere else? (https://openxbox.org/discord)

Uncoupling so far:
XvdFile.cs
```cs
private bool CryptSectionXts(bool encrypt, byte[] key, uint headerId, ulong offset, ulong length)
{
    int numDataUnits = (int)(length + 0xFFF) / 0x1000;
    var headerIdBytes = BitConverter.GetBytes(headerId);

    var tweakAesKey = new byte[0x10];
    var dataAesKey = new byte[0x10];
    
    var tweak = new byte[0x10];

    // Split tweak- / Data AES key
    Array.Copy(key, tweakAesKey, 0x10);
    Array.Copy(key, 0x10, dataAesKey, 0, 0x10);

    // Copy VDUID and header Id as tweak
    Array.Copy(Header.VDUID, 0, tweak, 0x8, 0x8);
    Array.Copy(headerIdBytes, 0, tweak, 0x4, 0x4);

    var cipherXts = new AesXts(tweakAesKey, dataAesKey, tweak);

    _io.Stream.Position = (long)offset;
    for (int dataUnit = 0; dataUnit < numDataUnits; dataUnit++)
    {
        var origData = _io.Reader.ReadBytes(0x1000);
        _io.Stream.Position -= 0x1000;

        byte[] newData = cipherXts.CryptData(encrypt, origData, dataUnit);

        _io.Writer.Write(newData);
    }
    return true;
}
```

AesXts.cs
```cs
using System;
using System.IO;

namespace LibXboxOne
{
    public class AesXts
    {
        const int AES_BLOCK_SIZE = 0x10;
        readonly byte[] CounterAesKey;
        readonly byte[] DataAesKey;
        readonly byte[] InitialCounterBytes;

        readonly AesCipher CounterCipher;
        readonly AesCipher DataCipher;

        public AesXts(byte[] counterAesKey, byte[] dataAesKey, byte[] initialCounterBytes)
        {
            CounterAesKey = counterAesKey;
            DataAesKey = dataAesKey;
            InitialCounterBytes = initialCounterBytes;

            CounterCipher = new AesCipher(counterAesKey);
            DataCipher = new AesCipher(DataAesKey);
        }

        public static byte[] MultiplyTweak(byte[] tweak)
        {
            byte dl = 0;
            var newTweak = new byte[0x10];

            for (int i = 0; i < 0x10; i++)
            {
                byte cl = tweak[i];
                byte al = cl;
                al = (byte)(al + al);
                al = (byte)(al | dl);
                dl = cl;
                newTweak[i] = al;
                dl = (byte)(dl >> 7);
            }
            if (dl != 0)
                newTweak[0] = (byte)(newTweak[0] ^ 0x87);
            return newTweak;
        }
        public byte[] CryptData(bool encrypt, byte[] data, int dataUnit)
        {
            int blocks = data.Length / AES_BLOCK_SIZE;
            var newData = new byte[data.Length];
            byte[] encryptedTweak = new byte[0x10];
            byte[] tweak = InitialCounterBytes;

            // Update tweak with data unit number
            Array.Copy(BitConverter.GetBytes(dataUnit), tweak, 4);

            // Encrypt tweak
            CounterCipher.EncryptBlock(tweak, 0, tweak.Length, encryptedTweak, 0);

            for (int i = 0; i < blocks; i++)
            {
                // Encrypt data, using encrypted tweak as IV
                byte[] crypted = CryptBlock(encrypt, data, i * AES_BLOCK_SIZE, AES_BLOCK_SIZE, encryptedTweak);
                Array.Copy(crypted, 0, newData, i * AES_BLOCK_SIZE, AES_BLOCK_SIZE);

                encryptedTweak = MultiplyTweak(encryptedTweak);
            }
            return newData;
        }

        byte[] CryptBlock(bool encrypt, byte[] data, int dataOffset, int dataLength, byte[] encryptedTweak)
        {
            var newData = new byte[dataLength];

            //if (!encrypt)
            for (int i = 0; i < dataLength; i++)
            {
                newData[i] = (byte)(data[dataOffset + i] ^ encryptedTweak[i % encryptedTweak.Length]);
            }

            var cryptData = new byte[dataLength];

            if(encrypt)
                DataCipher.EncryptBlock(newData, 0, dataLength, cryptData, 0);
            else
                DataCipher.DecryptBlock(newData, 0, dataLength, cryptData, 0);

            for (int i = 0; i < dataLength; i++)
            {
                cryptData[i] = (byte)(cryptData[i] ^ encryptedTweak[i % encryptedTweak.Length]);
            }

            return cryptData;
        }
    }
}
```

**Comment by tuxuser** [comment]
ThirdParty/AesCipher.cs can be replaced by simply using "RSA - ECB - NoPadding"

**Comment by tuxuser** [comment]
**NOTE**: In this implementation, SHA256 / RSA signature verification functions take the data instead of a precomputed hash!

To get rid of NCrypt:

BCryptRsaImport.cs
```cs
using System;
using System.IO;
using System.Runtime.InteropServices;
using System.Security.Cryptography;

namespace LibXboxOne
{
    public enum BCRYPT_RSABLOB_MAGIC : uint
    {
        // The key is an RSA public key. 
        RSAPUBLIC = 0x31415352,
        // The key is an RSA private key. 
        RSAPRIVATE = 0x32415352,
        // The key is a full RSA private key. 
        RSAFULLPRIVATE = 0x33415352
    };

    [StructLayout(LayoutKind.Sequential)]
    public struct BCRYPT_RSAKEY_BLOB
    {
        public BCRYPT_RSABLOB_MAGIC Magic;
        public uint BitLength;
        public uint cbPublicExp;
        public uint cbModulus;
        public uint cbPrime1;
        public uint cbPrime2;
    };

    public sealed class BCryptRsaImport
    {
        public static RSAParameters BlobToParameters(byte[] blobData)
        {
            var parameters = new RSAParameters();
            var reader = new BinaryReader(new MemoryStream(blobData));

            BCRYPT_RSAKEY_BLOB header = reader.ReadStruct<BCRYPT_RSAKEY_BLOB>();
            if (header.Magic != BCRYPT_RSABLOB_MAGIC.RSAPUBLIC &&
                header.Magic != BCRYPT_RSABLOB_MAGIC.RSAPRIVATE &&
                header.Magic != BCRYPT_RSABLOB_MAGIC.RSAFULLPRIVATE)
            {
                throw new InvalidDataException("Unexpected RSA keyblob");
            }

            parameters.Exponent = reader.ReadBytes((int)header.cbPublicExp);
            parameters.Modulus = reader.ReadBytes((int)header.cbModulus);

            if (header.Magic != BCRYPT_RSABLOB_MAGIC.RSAPUBLIC)
            {
                // RSAPRIVATEBLOB
                parameters.P = reader.ReadBytes((int)header.cbPrime1);
                parameters.Q = reader.ReadBytes((int)header.cbPrime2);

                if (header.Magic != BCRYPT_RSABLOB_MAGIC.RSAPRIVATE)
                {
                    // RSAFULLPRIVATEBLOB
                    parameters.DP = reader.ReadBytes((int)header.cbPrime1);
                    parameters.DQ = reader.ReadBytes((int)header.cbPrime2);
                    parameters.InverseQ = reader.ReadBytes((int)header.cbPrime1);
                    parameters.D = reader.ReadBytes((int)header.cbModulus);
                }
            }

            return parameters;
        }
    }
}
```

SHA.cs
```cs
using System;
using System.Security.Cryptography;

namespace LibXboxOne
{
    public static class SHA
    {
        public static byte[] ComputeSHA1(byte[] data, int offset, int length)
        {
            byte[] dataToHash = new byte[length];
            Array.Copy(data, offset, dataToHash, 0, length);
            return ComputeSHA1(dataToHash);
        }

        public static byte[] ComputeSHA1(byte[] data)
        {
            var hasher = SHA1.Create();
            return hasher.ComputeHash(data);
        }

        public static byte[] ComputeSHA256(byte[] data, int offset, int length)
        {
            byte[] dataToHash = new byte[length];
            Array.Copy(data, offset, dataToHash, 0, length);
            return ComputeSHA256(dataToHash);
        }

        public static byte[] ComputeSHA256(byte[] data)
        {
            var hasher = SHA256.Create();
            return hasher.ComputeHash(data);
        }

        public static void KeyAppendInt(ref byte[] key, uint appendData)
        {
            byte[] hash = ComputeSHA256(BitConverter.GetBytes(appendData));
            KeyAppend(ref key, hash);
        }

        public static void KeyAppend(ref byte[] key, byte[] appendData)
        {
            bool bit = false;
            for(int i = 0; i < 256; ++i)
            {
                bit = (appendData[31 - i / 8] & (1 << (i & 7))) > 0;
                byte[] data = new byte[0x20];
                data[0x20 - 1] = (byte)(bit ? 1 : 0);
                KeyAppendSha(ref key, data);
            }
        }

        public static void KeyAppendSha(ref byte[] key, byte[] appendData)
        {
            SHA256 hash = SHA256.Create();
            hash.TransformBlock(key, 0, key.Length, null, 0);
            hash.TransformFinalBlock(appendData, 0, appendData.Length);

            key = hash.Hash;
        }

        public static byte[] CalculateSignature_SHA256withRSA(byte[] rsaKeyBlobBCrypt, byte[] data)
        {
            var rsaParams = BCryptRsaImport.BlobToParameters(rsaKeyBlobBCrypt);

            RSA rsa = RSA.Create();
            rsa.ImportParameters(rsaParams);

            return rsa.SignData(data, HashAlgorithmName.SHA256, RSASignaturePadding.Pss);
        }

        // Return true if valid, false otherwise
        public static bool VerifySignature_SHA256withRSA(byte[] rsaKeyBlobBCrypt, byte[] signature, byte[] data)
        {
            var rsaParams = BCryptRsaImport.BlobToParameters(rsaKeyBlobBCrypt);

            RSA rsa = RSA.Create();
            rsa.ImportParameters(rsaParams);

            return rsa.VerifyData(data, signature, HashAlgorithmName.SHA256, RSASignaturePadding.Pss);
        }
    }
}
```

**Comment by emoose** [comment]
Nice work, maybe with AES replaced we can get a speed improvement too, IIRC it used to be pretty dreadful working with larger files.

Apparently AesCryptoServiceProvider can make use of AES-NI if CPU & OS support it (doesn't seem to extend to the easier to use AesManaged class though), but that could perform worse if using small inputs due to call-overhead (according to https://stackoverflow.com/questions/23444135/is-net-aesmanaged-cryptography-hardware-accelerated#comment36592512_23531131), and right now we only run AES over 16-byte blocks thanks to the XTS tweak etc, maybe there's a way to change it to use a larger buffer?

I'll try coming on Discord soon, add me if you get a chance: emoose#6289

**Comment by tuxuser** [comment]
Added you on Discord

XvdFile.cs
```cs
private ulong CalculateHashBlockNumForBlockNum(uint unk1, ulong blockNum, uint idx, out ulong entryNumInBlock)
{
    ulong xvdBlockCount = XvdDataBlockCount;
    long hashTreeLevels = (long)HashTreeLevels;
    entryNumInBlock = 0;

    if (unk1 > 1 || idx > 3)
        return 0xFFFF;
    else if (idx == 3)
        return 0;
    else if (idx == 0)
        entryNumInBlock = blockNum % 0xAA;
    else
        entryNumInBlock = blockNum / (ulong)Math.Pow(0xAA, idx) % 0xAA;

    ulong result = blockNum / (ulong)Math.Pow(0xAA, idx + 1);
    hashTreeLevels -= (idx + 1);

    ulong tmpSize;
    if ((idx == 0 || idx == 2) && hashTreeLevels > 0)
    {
        tmpSize = (ulong)Math.Pow(0xAA, idx + 2);
        result += (xvdBlockCount + tmpSize - 1) / tmpSize;
        if (idx == 0)
            hashTreeLevels--;
    }

    if ((idx == 0 || idx == 1) && hashTreeLevels > 0)
    {
        tmpSize = (ulong)Math.Pow(0xAA, 3);
        result += (xvdBlockCount + tmpSize - 1) / tmpSize;
        hashTreeLevels--;

        if (hashTreeLevels > 0)
        {
            tmpSize = (ulong)Math.Pow(0xAA, 4);
            result += (xvdBlockCount + tmpSize - 1) / tmpSize;
        }
    }

    return result;
}

public static ulong CalculateNumHashBlocksInLevel(ulong size, ulong idx)
{
    if (idx > 3)
        return 0;

    ulong blockSz = (ulong)Math.Pow(0xAA, idx + 1);
    return (size + blockSz - 1) / blockSz;
}
```

**Comment by tuxuser** [comment]
meh, hitting a roadblock here.
https://github.com/dotnet/corefx/blob/master/Documentation/architecture/cross-platform-cryptography.md#rsa

RSA with PSS is only supported by Windows via CNG crypto provider ...

**EDIT**: Had to utilize BouncyCastle to achieve signature implementation: https://github.com/tuxuser/xvdtool/blob/tests/LibXboxOne/Crypto/HashUtils.cs#L42

**Comment by tuxuser** [comment]
Solved by merging PR #10, thanks! 👍
